### ISSUE TITLE
SuperAgentWrapper observations from dead agents

### ISSUE THREAD

**Issue by rusu24edward** [LLNL/Abmarl#290]
The super agent wrapper generates observations for dead agents, giving them an advantage by having access to more information in the grid. The super agent wrapper needs to define null observations for each of its agents.

I don't think it needs to define null actions since it just removes those before giving it to the sim. We should add an example with the wrapper using the parametric action space.

We may have to define null rewards since the agent can still be rewarded/penalized indirectly.

**Comment by rusu24edward** [comment]
When an agent has just died in this step, we can report its null obs, or we can report its last observation.
1. If we report the null obs, then the "terminating state" for that agent will be the null obs. This may screw with the back-propagation during training since penultimate states will finish with the null state. When we end in a good state (e.g. some target or goal), we'll see the null obs, and when we end in bad state (e.g. dead), we'll see the null obs.
2. If we report the last obs, then we don't have the same null obs issue as above. Whether an agent reaches a "good" terminating state or a "bad" terminating state, we keep the transition information. On the other hand, the agent's observations may include changes that have occurred since its death. Suppose the agent dies part way through a step. Its last obs will include the total processing of that step, including passive dynamics from the simulation and the actions of all the other agents. Not until the next step (or the agent's next turn) do we gather the observation, so this may still give the agent an unfair advantage.

Furthermore, we should also consider the implications on the rewards. Reporting a reward of 0 for a dead agent doesn't make any sense because we often penalize the agent for dying, and it needs to receive that penalty. Furthermore, we often reward agents for accomplishing their goal (e.g. reaching a target), and we want the agent to see that too. After an agent is done, we should return zero rewards for future steps since the agent is no longer a part of the game, and thus anything that happens to it should not be considered. Importantly, we do need to consider the "in-between": when an agent is done, its reward may still be affected by the other agents' actions. Should we include that in the final reporting of the reward?

Regarding the observations, the repercussions from (1) are difficult to measure. We would be fixing the corrupt data by potentially corrupting it in a worse way. (2) is not perfect either because there may still be some unfair advantage, but only in that single step, since in all future steps the agent is done and should null. #78 suggests that the simulation takes a snapshot of the observation just as the agent dies, and we could report that as its final observation. This impacts the reporting of the final reward.

Obs reporting and reward reporting need to sync up. So whatever we decide for one, we have to decide for the other. In my opinion, it makes the most sense that the final reward an agent receives as it transitions to done is the reward that takes into account all the actions in the game. This is how all the other transitions work, so this is how the final transition should work. In this case, the final observation should be the observation that takes into account all the actions of the other agents, even if the observing agent died somewhere in there.

In conclusion, the final observation and reward should include all the dynamics that occur from the point of the done-ness to the actual reporting. All *future* obs and rewards should be null. We may be able to implement this in the SuperAgentWrapper the same way we do in a Manager: by keeping track of our own set of done agents so that we can identify when an agent has just finished.

These notes should be considered for #279

**Comment by rusu24edward** [comment]
Normally, `get_{obs, reward, done, info}` do not impact the state of the simulation. This should be true for a simulation since these functions simply query the state and should not change it. However, it is not necessarily true for the wrappers, and if the SuperAgentWrapper is responsible for tracking the done state of the agents, then we break the assumption that the getters only query the sim and don't modify it. That becomes part of the meta-state, which only the wrapper uses. However, since the wrapper is treated liked a AgentBasedSimulation by the rest of the framework, we would have state-changing query functions.

If we go down this route, then we can create a `query` function in the AgentBasedSimulation which calls the four getters *in order*: obs, reward, done info. The managers can use this query function instead of calling the functions individually. This is important because we have to ask, "At what point in the SuperAgentWrapper" do we store the agent as done? At `get_done`? Sure, but then that means that we assume that the Manager will always call `get_reward` before it calls `get_done`.

**Note**: Technically, `get_reward` does change the state by resetting the reward to 0, but this practice is assumed already and doesn't assume anything about how the other getters are used.

**Comment by rusu24edward** [comment]
> then we break the assumption that the getters only query the sim and don't modify it.

I think I can get around this by putting the logic in step, which is safe since step must *always* be called before a getter.

**Comment by rusu24edward** [comment]
> I think I can get around this by putting the logic in step, which is safe since step must _always_ be called before a getter.

Suppose I have `super0: (agent0, agent1)` and `agent2`. Now suppose the following sequence of events:
1. `agent2` takes an action, causing `agent0` to die.
2. Observation call for `super0`. Because `agent0` has not acted, it will give normal observation
3. `super0` acts, `agent0` moved to the `just_died` set. It should have already been there.
4. Observation call for `super0`. `agent0` will give an observation again.
5. `super0` acts, and `agent0` is moved to the `already_dead` set. From here on it will give null obs.
From this sequence, we see that `agent0` produced one too many observations.

If we put the set processing outside the loop, then we can find a scenario where `agent0` dies and if `agent1` acts again, it will be moved from `just_died` to `already_dead` before producing its last observation.

Thus, we must put the processing in `get_obs`. But get reward depends on the sets too, and we don't know if obs was called before or after reward since we don't impose an order. The `query` function would solve this, but I don't think this is a very good design...

**Comment by rusu24edward** [comment]
I can make the managers responsible for deciding when to call the null obs or not. Then I can add a null obs property to the ObservingAgent class. If it is defined, then use that. The Observers in GSF can give the agents their null obs.

**Comment by rusu24edward** [comment]
> I can make the managers responsible for deciding when to call the null obs or not.

Managers can't access covered agents, so this has to happen in the SuperAgentWrapper.

**Comment by rusu24edward** [comment]
Just have separate variables for checking the last report status for the obs and the reward.
